apollo-runtime/src/main/java/com/apollographql/apollo/subscription/WebSocketSubscriptionTransport.kt

```diff
@@ -0,0 +1,165 @@
+      delegateRef.get()?.onMessage(OperationServerMessage.fromJsonString(text))
+    }
+
+    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+      delegateRef.get()?.onFailure(t)
+    }
+
+    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+      delegateRef.get()?.onClosed()
+    }
+
+    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+      delegateRef.get()?.onClosed()
+    }
+
+    fun release() {
+      delegateRef.clear()
+    }
+  }
+
+  /**
+   * @param webSocketUrl The URL of the GraphQL WebSocket API
+   * @param webSocketConnectionFactory The [WebSocket.Factory] to use to create the websockets
+   * @param writePayloadAsJsonString If the payload should be serialized as a JSON String (default false). Set this to `true` for AWS
+   *                                 AppSync integrations.
+   * @param extensions Any additional extensions to when sending start commands. Must not be changed after this creating the factory.
+   */
+  class Factory @JvmOverloads constructor(
+      webSocketUrl: String,
+      private val webSocketConnectionFactory: WebSocket.Factory,
+      private val writePayloadAsJsonString: Boolean = false,
+      private val extensions: Map<String, Any?> = emptyMap()
+  ) : SubscriptionTransport.Factory {
+    private val webSocketRequest: Request = Request.Builder()
+        .url(webSocketUrl)
+        .addHeader("Sec-WebSocket-Protocol", "graphql-ws")
+        .addHeader("Cookie", "")
+        .build()
+
+    override fun create(callback: SubscriptionTransport.Callback): SubscriptionTransport =
+        WebSocketSubscriptionTransport(webSocketRequest, webSocketConnectionFactory, callback, writePayloadAsJsonString, extensions)
+  }
+
+  class AppSyncFactory @JvmOverloads constructor(
+      webSocketUrl: String,
+      private val webSocketConnectionFactory: WebSocket.Factory,
+      authorization: Map<String, Any?>,
+      payload: Map<String, Any?> = emptyMap()
+  ) : SubscriptionTransport.Factory by Factory(
+      webSocketUrl = webSocketUrl
+          .let {
+            HttpUrl.get(when {
+              it.startsWith("ws://", ignoreCase = true) -> "http" + it.drop(2)
+              it.startsWith("wss://", ignoreCase = true) -> "https" + it.drop(3)
+              else -> it
+            })
+          }
+          .newBuilder()
+          .setQueryParameter("header", authorization.encodeAsQueryParam())
+          .setQueryParameter("payload", payload.encodeAsQueryParam())
+          .build()
+          .toString(),
+      webSocketConnectionFactory = webSocketConnectionFactory,
+      writePayloadAsJsonString = true,
+      extensions = mapOf("authorization" to authorization)
+  ) {
+    companion object {
+      private fun Map<String, Any?>.encodeAsQueryParam(): String {
+        val buffer = Buffer()
+        JsonWriter.of(buffer).jsonValue(this)
+        return Base64.getUrlEncoder().encodeToString(buffer.readByteArray())
```

Comment:
Maybe call this method `toBase64()` ? It's used as a query param but the method itself uses Base64 and not urlencoding.

Reply:
Done.